File: Quartz.Core/src/collections/IntMap.cpp
#include "IntMap.h"
 
void IntMap::set(IntInt element) {
    if (isEven(element.key)) even.set(element);
    else odd.set(element);
}

void IntMap::remove(uint32_t key) {
    if (isEven(key)) even.remove(key);
    else odd.remove(key);
}

uint32_t IntMap::tryGetValue(uint32_t key) const {
    if (isEven(key)) return even.tryGetValue(key);
    return odd.tryGetValue(key);
}

void IntMap::set(uint32_t key, uint32_t val) { set(IntInt{key, val});}

void IntMap::clear() {
    even.elements.clear();
    odd.elements.clear();
}

File: Quartz.CoreCs/src/native/collections/IntMap.cs
namespace Quartz.CoreCs.native.collections; 

public class IntMap : IDisposable {
	public unsafe void* ptr;
	public unsafe int count => QuartzNative.IntMap_Count(ptr);

	public unsafe IntMap() => ptr = QuartzNative.IntMap_Create();

	public unsafe void Set(IntInt element) => QuartzNative.IntMap_Set(ptr, element);
	public unsafe void Set(uint key, uint val) => QuartzNative.IntMap_Set(ptr, new(key, val));
	public unsafe void Remove(uint key) => QuartzNative.IntMap_Remove(ptr, key);
	public unsafe uint TryGet(uint key) => QuartzNative.IntMap_TryGetValue(ptr, key);
	public bool Contains(uint key) => TryGet(key) != uint.MaxValue;

	public unsafe void Clear() => QuartzNative.IntMap_Clear(ptr);

	public uint this[uint key] {
		get => TryGet(key);
		set => Set(key, value);
	}

	private unsafe void ReleaseUnmanagedResources() {
		if (ptr != null) QuartzNative.IntMap_Destroy(ptr);
		ptr = null;
	}
	
	public void Dispose() {
		ReleaseUnmanagedResources();
		GC.SuppressFinalize(this);
	}
	~IntMap() => ReleaseUnmanagedResources();
}